pkg/rdt/rdt.go

```diff
@@ -258,64 +274,93 @@ func (c *control) configureResctrl(conf config) error {
 	c.DebugBlock("", "applying resolved config: |\n%s", utils.DumpJSON(conf))
 
 	// Remove stale resctrl groups
-	existingClasses, err := c.classesFromResctrlFs()
+	classesFromFs, err := c.classesFromResctrlFs()
 	if err != nil {
 		return err
 	}
 
-	for _, cls := range existingClasses {
-		if _, ok := conf.Classes[cls.name]; !ok {
+	for name, cls := range classesFromFs {
+		if _, ok := conf.Classes[cls.name]; cls.name != RootClassName && !ok {
 			tasks, err := cls.GetPids()
 			if err != nil {
 				return rdtError("failed to get resctrl group tasks: %v", err)
 			}
 			if len(tasks) > 0 {
 				return rdtError("refusing to remove non-empty resctrl group %q", cls.relPath(""))
 			}
-			err = os.Remove(cls.path(""))
+			log.Debug("removing existing resctrl group %q", cls.relPath(""))
+			err = groupRemoveFunc(cls.path(""))
 			if err != nil {
 				return rdtError("failed to remove resctrl group %q: %v", cls.relPath(""), err)
 			}
+
+			delete(c.classes, name)
 		}
 	}
 
-	// Start with fresh set of classes. Root class is always present
-	c.classes = make(map[string]*ctrlGroup, len(conf.Classes))
-	c.classes[RootClassName], err = newCtrlGroup(c.resctrlGroupPrefix, RootClassName)
-	if err != nil {
-		return err
+	for name, cls := range c.classes {
+		if _, ok := conf.Classes[cls.name]; cls.name != RootClassName && !ok {
+			log.Debug("dropping stale class %q", name)
+			delete(c.classes, name)
+		}
+	}
+
+	if _, ok := c.classes[RootClassName]; !ok {
+		log.Panic("root class missing from runtime data, re-adding...")
+		c.classes[RootClassName] = classesFromFs[RootClassName]
```

Comment:
This still does not make sense to me. I don't see how this piece/line of code could ever be executed since the previous call to log.Panic() will `panic()`. IIUC, even if this was a `recover()`'ed `defer`red goroutine, that line would never be reached.